From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: qmuntal <qmuntaldiaz@microsoft.com>
Date: Mon, 30 Sep 2024 14:41:28 +0200
Subject: [PATCH] Support curve P-521 when TLS fipsonly mode is enabled

We have historically supported P-521 in TLS when fipsonly mode is
enabled, as this aligns with CNG, OpenSSL and SymCrypt FIPs policies.

Upstream follows the boringssl FIPS policy, which doesn't allow P-521.

This change adds support for P-521 in TLS when fipsonly mode is enabled.
---
 src/crypto/tls/defaults.go  | 3 ++-
 src/crypto/tls/fips_test.go | 3 ++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/crypto/tls/defaults.go b/src/crypto/tls/defaults.go
index f25d0d3ce3f9eb..08c091168d6383 100644
--- a/src/crypto/tls/defaults.go
+++ b/src/crypto/tls/defaults.go
@@ -102,7 +102,7 @@ var defaultSupportedVersionsFIPS = []uint16{
 
 // defaultCurvePreferencesFIPS are the FIPS-allowed curves,
 // in preference order (most preferable first).
-var defaultCurvePreferencesFIPS = []CurveID{CurveP256, CurveP384}
+var defaultCurvePreferencesFIPS = []CurveID{CurveP256, CurveP384, CurveP521}
 
 // defaultSupportedSignatureAlgorithmsFIPS currently are a subset of
 // defaultSupportedSignatureAlgorithms without Ed25519 and SHA-1.
@@ -115,6 +115,7 @@ var defaultSupportedSignatureAlgorithmsFIPS = []SignatureScheme{
 	PKCS1WithSHA384,
 	ECDSAWithP384AndSHA384,
 	PKCS1WithSHA512,
+	ECDSAWithP521AndSHA512,
 }
 
 // defaultCipherSuitesFIPS are the FIPS-allowed cipher suites.
diff --git a/src/crypto/tls/fips_test.go b/src/crypto/tls/fips_test.go
index ab3c4465ddff0c..fc47e8ba8313cd 100644
--- a/src/crypto/tls/fips_test.go
+++ b/src/crypto/tls/fips_test.go
@@ -111,7 +111,7 @@ func isFIPSCipherSuite(id uint16) bool {
 
 func isFIPSCurve(id CurveID) bool {
 	switch id {
-	case CurveP256, CurveP384:
+	case CurveP256, CurveP384, CurveP521:
 		return true
 	}
 	return false
@@ -135,6 +135,7 @@ func isFIPSSignatureScheme(alg SignatureScheme) bool {
 		PKCS1WithSHA384,
 		ECDSAWithP384AndSHA384,
 		PKCS1WithSHA512,
+		ECDSAWithP521AndSHA512,
 		PSSWithSHA256,
 		PSSWithSHA384,
 		PSSWithSHA512:
